### PR TITLE
rpc: In `utxoupdatepsbt` also look for the tx in the txindex

### PR DESCRIPTION
**Pull request by ishaanam**
Previously the `utxoupdatepsbt` RPC, added in #13932, only updated the inputs spending segwit outputs with the `witness_utxo`, because the full transaction is needed for legacy inputs. Before this RPC looked for just the utxos in the utxo set and the mempool. This PR makes it so that if the user has txindex enabled, then the full transaction is looked for there for all inputs. If it is not found in the txindex or  txindex isn't enabled, then the mempool is checked for the full transaction. If the transaction an input is spending from is still not found at that point, then the utxo set is searched and the inputs spending segwit outputs are updated with just the utxo.

### CODE REVIEW

**Review by w0xlt — COMMENTED** [review]
Approach ACK.
If the txindex is enabled and updated, it should be used.

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -172,44 +172,68 @@ PartiallySignedTransaction ProcessPSBT(const std::string& psbt_string, const std
...
 
-        const Coin& coin = view.AccessCoin(psbtx.tx->vin[i].prevout);
+    // If we still haven't found all of the inputs, look for the missing ones in the utxo set
+    if (!coins.empty()) {
+        FindCoins(node, coins);
+        for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
+            PSBTInput& input = psbtx.inputs.at(i);
+
+            // If there are still missing utxos, add them if they were found in the utxo set
+            if (!coins.empty() && !input.non_witness_utxo) {
```
Comment (achow101) [inline src/rpc/rawtransaction.cpp:216]: In 687d6c20a4d6fc99cf4ae5794dd4203ea446a8f2 "rpc, tests: in `utxoupdatepsbt` also look for the transaction in the txindex"

Checking `!coins.empty()` again is redundant. Also, we may still want to add the witness_utxo even if a non_witness_utxo exists.

We add the different utxo types depending on the script type. For legacy scripts, only non_witness_utxo should be added. For Segwit v0, both non_witness_utxo and witness_utxo should be added, For Segwit v1+, only witness_utxo should be added. This RPC should (try) to adhere to these guidelines.

Comment (ishaanam) [reply src/rpc/rawtransaction.cpp]: > Also, we may still want to add the witness_utxo even if a non_witness_utxo exists.

If the `non_witness_utxo` exists and a segwit output is being spent, `SignPSBTInput` will add the `witness_utxo`.
> For Segwit v1+, only witness_utxo should be added.

I have updated this commit to adhere to this, but this [comment](https://github.com/bitcoin/bitcoin/blob/master/src/psbt.cpp#L417-L419) indicates something a bit different: 
> We can remove the non_witness_utxo if and only if there are no non-segwit or segwit v0
    inputs in this transaction.

So does the `non_witness_utxo` need to be present for a Segwit v1 input if there is another non-segwit input int he transaction?

Comment (ishaanam) [reply src/rpc/rawtransaction.cpp]: > I think if there is at least one witv0 or legacy input in a transaction, the non_witness_utxo record of all inputs should be kept.

This is also what `walletprocesspsbt` does, so I have update this PR to do that as well.

**Comment by achow101** [comment]
ACK f26d600d55d86099206456f394a422436d5f3ec3

**Comment by ishaanam** [comment]
Rebased

**Comment by achow101** [comment]
ACK 2ee94290bdbb66732f420dc9d74493436fb2a59f

File: src/psbt.h [hunk 2]
```diff
@@ -1231,6 +1231,9 @@ bool PSBTInputSignedAndVerified(const PartiallySignedTransaction psbt, unsigned
  **/
 bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, const PrecomputedTransactionData* txdata, int sighash = SIGHASH_ALL, SignatureData* out_sigdata = nullptr, bool finalize = true);
 
+/** Removes the unnecessary non_witness_utxos from a psbt. */
```
Comment (Xekyo) [inline src/psbt.h:1234]: I see that you’re just moving this code, but could you add a little more explanation here? When I was reviewing this, I was first trying to find out what `non_witness_utxos` and `witness_utxos` were, when we needed them, and finally learned per the previous PR when we could drop them. My understanding is that we can only drop `non_witness_utxos` when all inputs are segwit v1.

I would suspect that we might also be able to drop `witness_utxos` on non-segwit inputs when we do have the corresponding `non_witness_utxos`, and vice versa, drop the `non_witness_utxos` on segwit inputs, if we do have the `witness_utxos` for them. Is that accurate?

```suggestion
/** Reduces the size of the PSBT by dropping unnecessary `non_witness_utxos` (i.e. previous transactions) from a psbt when all inputs are segwit v1. */
```

**Review by murchandamus — COMMENTED** [review]
ACK 2ee94290bdbb66732f420dc9d74493436fb2a59f

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -179,50 +179,77 @@ PartiallySignedTransaction ProcessPSBT(const std::string& psbt_string, const std
...
-    // Fill the inputs
-    const PrecomputedTransactionData txdata = PrecomputePSBTData(psbtx);
     for (unsigned int i = 0; i < psbtx.tx->vin.size(); ++i) {
-        PSBTInput& input = psbtx.inputs.at(i);
+        PSBTInput& psbt_input = psbtx.inputs.at(i);
+        const CTxIn& tx_in = psbtx.tx->vin.at(i);
 
-        if (input.non_witness_utxo || !input.witness_utxo.IsNull()) {
-            continue;
+        if (psbt_input.non_witness_utxo) continue;
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/rpc/rawtransaction.cpp:197]: Done

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -179,50 +179,77 @@ PartiallySignedTransaction ProcessPSBT(const std::string& psbt_string, const std
...
-        LOCK2(cs_main, mempool.cs);
-        CCoinsViewCache &viewChain = chainman.ActiveChainstate().CoinsTip();
-        CCoinsViewMemPool viewMempool(&viewChain, mempool);
-        view.SetBackend(viewMempool); // temporarily switch cache backend to db+mempool view
-
-        for (const CTxIn& txin : psbtx.tx->vin) {
-            view.AccessCoin(txin.prevout); // Load entries from viewChain into view; can fail.
-        }
+    // Fetch previous transactions:
+    // First, look in the txindex and the mempool
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/rpc/rawtransaction.cpp]: Yes, I've moved it.

File: src/psbt.h (hunk 2 above)
Comment (ishaanam) [reply src/psbt.h]: > I see that you’re just moving this code, but could you add a little more explanation here?

Done as per your suggestion above.
> I would suspect that we might also be able to drop witness_utxos on non-segwit inputs when we do have the corresponding non_witness_utxos,

I don't think `witness_utxo`s should ever get added to non-segwit inputs. 
> and vice versa, drop the non_witness_utxos on segwit inputs, if we do have the witness_utxos for them.

Currently I'm not sure, though in https://github.com/bitcoin/bitcoin/pull/25939#discussion_r1020917822 sipa said that in this situation all `non_witness_utxo`s should be kept. Even if this is not the case, changing this behavior would be out of the scope of this PR.

**Comment by Xekyo** [comment]
ACK 6e9f8bb050785dbc754b6bb493aad9139908ef98

**Pull request merged**